### PR TITLE
Return if column is generated or not

### PR DESCRIPTION
**Pull request by devotis**
Related to
https://github.com/strongloop/loopback-datasource-juggler/issues/899

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by superkhau** [comment]
Can you add some tests to verify your changes and prevent regressions in the future?

**Comment by devotis** [comment]
@superkhau I've added tests

**Comment by devotis** [comment]
@superkhau Can you help me with the unsuccessful check: PR out of date, needs to be rebased? I'm not sure how that happened.

**Comment by superkhau** [comment]
Something is weird about this PR, you're getting our translation commits mixed in there for some reason -- can you rebase again before I add my code review comments?

**Comment by superkhau** [comment]
@devotis It means other commit have landed on master and your pull request needs to get those changes. Do some googling for `git rebase` for more info.

**Comment by devotis** [comment]
I just rebased, but whenever I push here, I keep seeing commits that aren't mine. I do not understand why.

I'm on `feature-generated` of my fork. This is what I did.
- `git fetch upstream`
- `git rebase upstream/master`, with output

```
First, rewinding head to replay your work on top of it...
Applying: Return if column is generated or not
Applying: Tests for generated columns
Applying: Update translation files - round#2
Using index info to reconstruct a base tree...
M	intl/es/messages.json
M	intl/ja/messages.json
M	intl/tr/messages.json
<stdin>:11: trailing whitespace.
  "80a32e80cbed65eba2103201a7c94710": "No se ha encontrado el modelo: {0}",
<stdin>:24: trailing whitespace.
  "9dca910dae94ac66098b3992911e1b23": "{{Microsoft SQL Server}} では正規表現演算子はサポートされません。"
<stdin>:35: trailing whitespace.
  "a0078d732b2dbabf98ed2efcdb55b402": "{{table}} zorunlu bir dizgi bağımsız değişkeni: {0}",
warning: 3 lines add whitespace errors.
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.
Applying: Add connectorCapabilities global object (#102)
Using index info to reconstruct a base tree...
M	test/init.js
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.
Applying: Return if column is generated or not
Using index info to reconstruct a base tree...
M	lib/discovery.js
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.
Applying: Tests for generated columns
Using index info to reconstruct a base tree...
M	test/discover.test.js
M	test/tables.sql
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.
Applying: Fix a few dumb typos
```

- Then `git status` tells me:
```
On branch feature-generated
Your branch and 'origin/feature-generated' have diverged,
and have 4 and 9 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
nothing to commit, working directory clean
```

- Then `git push` tells me I need need to pull first, which I don't understand, because I was in sync with my remote before I rebased.
```
To https://github.com/devotis/loopback-connector-mssql.git
 ! [rejected]        feature-generated -> feature-generated (non-fast-forward)
error: failed to push some refs to 'https://github.com/devotis/loopback-connector-mssql.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
- So I do `git pull` first
```
First, rewinding head to replay your work on top of it...
```
- Followed by a `git push`
```
Everything up-to-date
```
And I still see the commits in the PR that aren't mine.

Maybe a few days before I merged origin/master into my branch, but I'm not sure If that's what I did. Maybe I need to merge origin/master again? 

These are my remotes:
- origin	https://github.com/devotis/loopback-connector-mssql.git (fetch/push)
- strongloop	https://github.com/strongloop/loopback-connector-mssql.git (fetch/push)
- upstream	https://github.com/strongloop/loopback-connector-mssql.git (fetch/push)

**Comment by devotis** [comment]
@superkhau Forget about my remarks, I had to force push. This PR is now rebased!

File: test/discover.test.js [hunk 1]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
   });
 });
 
+describe('Discover model generated columns', function() {
+  it('should return an array of columns for product and none of them is generated', function(done) {
+    db.discoverModelProperties('product', function(err, models) {
+      if (err) {
```
Comment (superkhau) [inline test/discover.test.js]: if (err) return done(err);

File: test/discover.test.js [hunk 2]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+        console.error(err);
+        done(err);
+      } else {
```
Comment (superkhau) [inline test/discover.test.js]: remove the else

File: test/discover.test.js [hunk 3]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+        models.forEach(function(m) {
+          // console.dir(m);
```
Comment (superkhau) [inline test/discover.test.js]: remove comment

File: test/discover.test.js [hunk 4]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+          assert(m.tableName === 'product');
+          assert(!m.generated, 'product table should not have generated (identity) columns');
+
```
Comment (superkhau) [inline test/discover.test.js]: remove empty space

File: test/discover.test.js [hunk 5]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+        });
+        done(null, models);
```
Comment (superkhau) [inline test/discover.test.js]: done();

File: test/discover.test.js [hunk 6]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+describe('Discover model generated columns', function() {
+  it('should return an array of columns for product and none of them is generated', function(done) {
+    db.discoverModelProperties('product', function(err, models) {
+      if (err) {
+        console.error(err);
+        done(err);
+      } else {
+        models.forEach(function(m) {
+          // console.dir(m);
+          assert(m.tableName === 'product');
```
Comment (superkhau) [inline test/discover.test.js]: assert.equal(m.tableName, 'product');

File: test/discover.test.js [hunk 7]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+          assert(!m.generated, 'product table should not have generated (identity) columns');
+
+        });
+        done(null, models);
+      }
+    });
+  });
```
Comment (superkhau) [inline test/discover.test.js]: add empty line to space out tests

File: test/discover.test.js [hunk 8]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+  it('should return an array of columns for testgen and the first is generated', function(done) {
+    db.discoverModelProperties('testgen', function(err, models) {
+      if (err) {
```
Comment (superkhau) [inline test/discover.test.js]: if (err) return done(err);

File: test/discover.test.js [hunk 9]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+        console.error(err);
+        done(err);
+      } else {
```
Comment (superkhau) [inline test/discover.test.js]: remove else

File: test/discover.test.js [hunk 10]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+        models.forEach(function(m) {
+          // console.dir(m);
```
Comment (superkhau) [inline test/discover.test.js]: remove comment

File: test/discover.test.js [hunk 11]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+          assert(m.tableName === 'testgen');
```
Comment (superkhau) [inline test/discover.test.js]: assert.equal(m.tableName, 'testgen');

File: test/discover.test.js [hunk 12]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+          if (m.columnName === 'id') {
+            assert(m.generated, 'testgen.id should be a generated (identity) column');
+          }
+        });
+        done(null, models);
```
Comment (superkhau) [inline test/discover.test.js]: done();

File: test/discover.test.js [hunk 13]
```diff
@@ -169,6 +169,43 @@ describe('Discover model foreign keys', function() {
...
+      }
+    });
+  });
+});
+
+
```
Comment (superkhau) [inline test/discover.test.js]: remove extra empty line

**Comment by superkhau** [comment]
GJ figuring the rebase BTW. :+1:

**Comment by loay** [comment]
@slnode test please

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by loay** [comment]
@devotis Can you please rebase? Thanks.

**Comment by devotis** [comment]
I believe I just rebased. I don't really care about the commit linting problem. I'm happy to let somebody else fix that. [Edits are allowed by maintainers](https://github.com/blog/2247-improving-collaboration-with-forks). If that's not enough I can add somebody as a collaborator of my fork.

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@devotis Rebased your branch and fixed the commit linter.

File: test/discover.test.js [hunk 14]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
       }
     });
   });
+
+  it('should return just the primary keys and not the foreign keys for inventory', function(done) {
+    db.discoverPrimaryKeys('inventory', function(err, models) {
+      if (err) {
```
Comment (ssh24) [inline test/discover.test.js]: `if (err) return done(err);`

File: test/discover.test.js [hunk 15]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+          assert(m.pkName.match(/_pk$/));
+        });
+        done(null, models);
+      }
+    });
+  });
+
+  it('should return just the primary keys and not the foreign keys for dbo.inventory', function(done) {
+    db.discoverPrimaryKeys('inventory', {owner: 'dbo'}, function(err, models) {
+      if (err) {
```
Comment (ssh24) [inline test/discover.test.js]: `if (err) return done(err);`

File: test/discover.test.js [hunk 16]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+        console.error(err);
+        done(err);
+      } else {
+        models.forEach(function(m) {
+          // console.dir(m);
```
Comment (ssh24) [inline test/discover.test.js]: Remove not needed comment

File: test/discover.test.js [hunk 17]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+  });
+  it('should return the primary key for version which consists of two columns', function(done) {
+    db.discoverPrimaryKeys('version', function(err, models) {
+      if (err) {
+        console.error(err);
+        done(err);
+      } else {
+        assert(models.length === 2);
+        models.forEach(function(m) {
+          // console.dir(m);
```
Comment (ssh24) [inline test/discover.test.js]: Remove not needed comment

File: test/discover.test.js [hunk 18]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+          assert(m.tableName === 'version');
+          assert(m.pkName.match(/_pk$/));
+        });
+        done(null, models);
+      }
+    });
+  });
+  it('should return the primary key for dbo.version which consists of two columns', function(done) {
+    db.discoverPrimaryKeys('version', {owner: 'dbo'}, function(err, models) {
+      if (err) {
```
Comment (ssh24) [inline test/discover.test.js]: `if (err) return done(err);`

File: test/discover.test.js [hunk 19]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+        console.error(err);
+        done(err);
+      } else {
```
Comment (ssh24) [inline test/discover.test.js]: Remove else

File: test/tables.sql [hunk 20]
```diff
@@ -69,6 +79,17 @@ go
 	ttl integer
    ) ;
 
+<<<<<<< HEAD
```
Comment (ssh24) [inline test/tables.sql]: Fix this. Merge conflict error

File: test/discover.test.js [hunk 21]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
       }
     });
   });
+
+  it('should return just the primary keys and not the foreign keys for inventory', function(done) {
+    db.discoverPrimaryKeys('inventory', function(err, models) {
+      if (err) {
+        console.error(err);
+        done(err);
+      } else {
```
Comment (ssh24) [inline test/discover.test.js]: remove else

File: test/discover.test.js [hunk 22]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+        models.forEach(function(m) {
+          // console.dir(m);
+          assert(m.tableName === 'inventory');
```
Comment (ssh24) [inline test/discover.test.js]: use `assert.equal(m.tableName, 'inventory');` instead

File: test/discover.test.js [hunk 23]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+          assert(m.columnName === 'id');
```
Comment (ssh24) [inline test/discover.test.js]: use `assert.equal(m. columnName, 'id');` instead

File: test/discover.test.js [hunk 24]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+          assert(m.pkName.match(/_pk$/));
+        });
+        done(null, models);
```
Comment (ssh24) [inline test/discover.test.js]: just `done()` should be fine.

File: test/discover.test.js [hunk 25]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+      }
+    });
+  });
+
+  it('should return just the primary keys and not the foreign keys for dbo.inventory', function(done) {
+    db.discoverPrimaryKeys('inventory', {owner: 'dbo'}, function(err, models) {
+      if (err) {
+        console.error(err);
+        done(err);
+      } else {
```
Comment (ssh24) [inline test/discover.test.js]: remove else

File: test/discover.test.js [hunk 26]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+        models.forEach(function(m) {
+          // console.dir(m);
+          assert(m.tableName === 'inventory');
```
Comment (ssh24) [inline test/discover.test.js]: use `assert.equal(m.tableName, 'inventory');` instead

File: test/discover.test.js [hunk 27]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+          assert(m.columnName === 'id');
```
Comment (ssh24) [inline test/discover.test.js]: use `assert.equal(m. columnName, 'id');` instead

File: test/discover.test.js [hunk 28]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+          assert(m.pkName === 'inventory_pk');
```
Comment (ssh24) [inline test/discover.test.js]: use `assert.equal(m.pkName, 'id');` instead

File: test/discover.test.js [hunk 29]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+        });
+        done(null, models);
```
Comment (ssh24) [inline test/discover.test.js]: just `done()`

File: test/discover.test.js [hunk 30]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+      }
+    });
+  });
+  it('should return the primary key for version which consists of two columns', function(done) {
+    db.discoverPrimaryKeys('version', function(err, models) {
+      if (err) {
```
Comment (ssh24) [inline test/discover.test.js]: `if (err) done(err)`

File: test/discover.test.js [hunk 31]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+        console.error(err);
+        done(err);
+      } else {
```
Comment (ssh24) [inline test/discover.test.js]: remove else

File: test/discover.test.js [hunk 32]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+        assert(models.length === 2);
```
Comment (ssh24) [inline test/discover.test.js]: `assert.equal(models.length, 2)`

File: test/discover.test.js [hunk 33]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+        models.forEach(function(m) {
+          // console.dir(m);
+          assert(m.tableName === 'version');
```
Comment (ssh24) [inline test/discover.test.js]: `assert.equal(m.tableName, 'version')`

File: test/discover.test.js [hunk 34]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+          assert(m.pkName.match(/_pk$/));
+        });
+        done(null, models);
```
Comment (ssh24) [inline test/discover.test.js]: `done()`

File: test/discover.test.js [hunk 35]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+      }
+    });
+  });
+  it('should return the primary key for dbo.version which consists of two columns', function(done) {
+    db.discoverPrimaryKeys('version', {owner: 'dbo'}, function(err, models) {
+      if (err) {
+        console.error(err);
+        done(err);
+      } else {
+        assert(models.length === 2);
```
Comment (ssh24) [inline test/discover.test.js]: `assert.equal(models.length, 2)`

File: test/discover.test.js [hunk 36]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+        models.forEach(function(m) {
+          // console.dir(m);
+          assert(m.tableName === 'version');
```
Comment (ssh24) [inline test/discover.test.js]: `assert.equal(m.tableName, 'version')`

File: test/discover.test.js [hunk 37]
```diff
@@ -147,6 +147,72 @@ describe('Discover model primary keys', function() {
...
+          assert(m.pkName.match(/_pk$/));
+        });
+        done(null, models);
```
Comment (ssh24) [inline test/discover.test.js]: `done()`

**Comment by devotis** [comment]
'Fixing' that felt like a waste of my time, but I'm done anyway.

**Comment by ssh24** [comment]
@devotis Thank you. Really appreciate your will to contribute for us! :)

**Comment by ssh24** [comment]
@slnode test please

**Comment by devotis** [comment]
Need to find time later and also need to run the tests on a local machine. Otherwise I need to trial and error through commit and push and await Jenkins?

**Comment by ssh24** [comment]
@devotis These are the tests that are failing on CI. All of them added by you:
```

  1) Discover model primary keys should return just the primary keys and not the foreign keys for inventory:
     Uncaught AssertionError: null == true
      at C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\test\discover.test.js:134:9
      at Array.forEach (native)
      at C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\test\discover.test.js:130:14
      at C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-datasource-juggler\lib\observer.js:172:22
      at doNotify (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-datasource-juggler\lib\observer.js:99:49)
      at MsSQL.ObserverMixin._notifyBaseObservers (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-datasource-juggler\lib\observer.js:122:5)
      at MsSQL.ObserverMixin.notifyObserversOf (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-datasource-juggler\lib\observer.js:97:8)
      at cbForWork (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-datasource-juggler\lib\observer.js:162:14)
      at C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-connector\lib\sql.js:642:7
      at innerCB (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\lib\mssql.js:131:17)
      at C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\mssql\lib\main.js:1588:20
      at Request.userCallback (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\mssql\lib\tedious.js:853:61)
      at Request.callback (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\request.js:33:27)
      at Connection.message (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\connection.js:1179:27)
      at Connection.dispatchEvent (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\connection.js:519:45)
      at MessageIO.<anonymous> (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\connection.js:439:23)
      at ReadablePacketStream.<anonymous> (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\message-io.js:92:15)
      at readableAddChunk (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_readable.js:217:18)
      at ReadablePacketStream.Readable.push (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_readable.js:176:10)
      at ReadablePacketStream.Transform.push (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_transform.js:123:32)
      at ReadablePacketStream._transform (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\message-io.js:57:16)
      at ReadablePacketStream.Transform._read (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_transform.js:159:10)
      at ReadablePacketStream.Transform._write (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_transform.js:147:83)
      at doWrite (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_writable.js:345:64)
      at writeOrBuffer (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_writable.js:334:5)
      at ReadablePacketStream.Writable.write (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_writable.js:271:11)
      at Socket.ondata (_stream_readable.js:542:20)
      at readableAddChunk (_stream_readable.js:153:18)
      at Socket.Readable.push (_stream_readable.js:111:10)
      at TCP.onread (net.js:537:20)

  2) Discover model primary keys should return just the primary keys and not the foreign keys for dbo.inventory:

      Uncaught AssertionError: 'PK__inventor__3213E83F72215ED7' == 'inventory_pk'
      + expected - actual

      -PK__inventor__3213E83F72215ED7
      +inventory_pk
      
      at C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\test\discover.test.js:147:16
      at Array.forEach (native)
      at C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\test\discover.test.js:143:14
      at C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-datasource-juggler\lib\observer.js:172:22
      at doNotify (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-datasource-juggler\lib\observer.js:99:49)
      at MsSQL.ObserverMixin._notifyBaseObservers (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-datasource-juggler\lib\observer.js:122:5)
      at MsSQL.ObserverMixin.notifyObserversOf (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-datasource-juggler\lib\observer.js:97:8)
      at cbForWork (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-datasource-juggler\lib\observer.js:162:14)
      at C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\loopback-connector\lib\sql.js:642:7
      at innerCB (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\lib\mssql.js:131:17)
      at C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\mssql\lib\main.js:1588:20
      at Request.userCallback (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\mssql\lib\tedious.js:853:61)
      at Request.callback (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\request.js:33:27)
      at Connection.message (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\connection.js:1179:27)
      at Connection.dispatchEvent (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\connection.js:519:45)
      at MessageIO.<anonymous> (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\connection.js:439:23)
      at ReadablePacketStream.<anonymous> (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\message-io.js:92:15)
      at readableAddChunk (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_readable.js:217:18)
      at ReadablePacketStream.Readable.push (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_readable.js:176:10)
      at ReadablePacketStream.Transform.push (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_transform.js:123:32)
      at ReadablePacketStream._transform (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\tedious\lib\message-io.js:57:16)
      at ReadablePacketStream.Transform._read (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_transform.js:159:10)
      at ReadablePacketStream.Transform._write (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_transform.js:147:83)
      at doWrite (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_writable.js:345:64)
      at writeOrBuffer (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_writable.js:334:5)
      at ReadablePacketStream.Writable.write (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\readable-stream\lib\_stream_writable.js:271:11)
      at Socket.ondata (_stream_readable.js:542:20)
      at readableAddChunk (_stream_readable.js:153:18)
      at Socket.Readable.push (_stream_readable.js:111:10)
      at TCP.onread (net.js:537:20)

  3) Discover model primary keys should return the primary key for version which consists of two columns:
     Error: timeout of 9999ms exceeded. Ensure the done() callback is being called in this test.
      at [object Object].<anonymous> (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\mocha\lib\runnable.js:226:19)

  4) Discover model primary keys should return the primary key for dbo.version which consists of two columns:
     Error: timeout of 9999ms exceeded. Ensure the done() callback is being called in this test.
      at [object Object].<anonymous> (C:\jenkins\workspace\loopback-connector-mssql\68dfec3a\node_modules\mocha\lib\runnable.js:226:19)
```

**Comment by devotis** [comment]
Ah I get it. Those are tests I added, but they are intended for another pull request of mine that fixes a problem where foreign keys are retrieved as well when just primary keys are requested. I'll see how this mixup happened and how to proceed going forward.

**Comment by ssh24** [comment]
@devotis Okay. Please add the suitable unit tests for this patch instead then.

**Comment by ssh24** [comment]
@devotis Any update on this PR? I would like to proceed with this PR.

**Comment by devotis** [comment]
The tables inside the test database are not generated according to `test/tables.sql`. Please update your test database using this script:

https://github.com/devotis/loopback-connector-mssql/blob/d81d352d2245d9d57aadedea47567c14eac9f702/test/tables.sql

And the failing tests should pass...

**Comment by ssh24** [comment]
@devotis No those four failing tests has nothing to do with this PR. You have created this branch of your master which is related to another PR you have open https://github.com/strongloop/loopback-connector-mssql/pull/99 and hence it has those tests on here as well. Could you please remove those 4 test cases?

**Comment by devotis** [comment]
That's A. work for me now and B. more work later again to resolve the merge conflict that is going to happen if I remove the 4 test cases now... The database needs to be updated for PR #99 anyway. So why bother to remove these tests and introduce more work?

**Comment by ssh24** [comment]
@devotis Well those 4 tests are currently failing on the downstream connectors because the logic behind it is in #99. We should clean this up first before it proceeds any further or move everything to one PR for example #99 and proceed with that one.

**Comment by devotis** [comment]
Why not do #99 first? That one solves a bug. #100 is an enhancement.

**Comment by ssh24** [comment]
@devotis Fixed the merge conflicts for you. PR should be good now.

@slnode test please

**Pull request merged**